public_html/wp-content/mu-plugins/camptix-tweaks/addons/ticket-types/ticket-types.php

```diff
@@ -0,0 +1,188 @@
+	if ( count( $match ) > 0 ) {
+		return current( $match );
+	}
+	return 'default';
+}
+
+/**
+ * Get the type of a given ticket.
+ *
+ * @return array
+ */
+function get_type( $ticket_id ) {
+	$type = get_post_meta( $ticket_id, META_KEY, true );
+	if ( ! $type ) {
+		$type = 'default';
+	}
+	$match = wp_list_filter( get_types(), array( 'slug' => $type ) );
+	return current( $match );
+}
+
+/**
+ * Get the slug for the type of a given ticket.
+ *
+ * @return string
+ */
+function get_type_slug( $ticket_id ) {
+	$type = get_type( $ticket_id );
+	return isset( $type['slug'] ) ? $type['slug'] : 'default';
+}
+
+/**
+ * Add the Ticket Types metabox to ticket editor.
+ */
+function add_types_meta_box() {
+	\add_meta_box( 'tix_ticket_types', __( 'Ticket Type', 'wordcamporg' ), __NAMESPACE__ . '\metabox_ticket_type', 'tix_ticket', 'normal', 'high' );
+}
+
+/**
+ * Metabox callback for ticket type.
+ */
+function metabox_ticket_type() {
+	global $camptix;
+	$purchased = $camptix->get_purchased_tickets_count( get_the_ID() );
+	$selected = get_type_slug( get_the_ID() );
+	$types = get_types();
+	?>
+	<p>
+		<?php esc_html_e( 'The type of ticket will determine how we set up the required questions. This cannot be changed once someone buys a ticket.', 'wordcamporg' ); ?>
+	</p>
+	<p>
+		<ul>
+		<?php foreach ( $types as $type ) : ?>
+			<li><strong><?php echo esc_html( $type['name'] ); ?>:</strong> <?php echo esc_html( $type['description'] ); ?></li>
+		<?php endforeach; ?>
+		</ul>
+	</p>
+	<p>
+		<label for="<?php echo esc_attr( META_KEY ); ?>"><?php esc_attr_e( 'Type:', 'wordcamporg' ); ?></label>
+		<select
+			id="<?php echo esc_attr( META_KEY ); ?>"
+			name="<?php echo esc_attr( META_KEY ); ?>"
+			<?php echo ( $purchased <= 0 ) ? '' : 'disabled="disabled"'; ?>
+		>
+			<?php foreach ( $types as $type ) : ?>
+				<option value="<?php echo esc_attr( $type['slug'] ); ?>" <?php selected( $selected, $type['slug'] ); ?>>
+					<?php echo esc_html( $type['name'] ); ?>
+				</option>
+			<?php endforeach; ?>
+		</select>
+	</p>
+
+	<?php if ( $purchased > 0 ) : ?>
+	<p>
+		<?php esc_html_e( 'You can not change the type because one or more tickets have already been purchased.', 'wordcamporg' ); ?>
+	</p>
+	<?php endif; ?>
+
+	<div class="clear"></div>
+	<?php
+}
+
+/**
+ * Callback on `save_post`, used to set our type meta.
+ *
+ * @param int $post_id
+ */
+function save_post( $post_id ) {
+	if ( ! is_admin() ) {
+		return;
+	}
+
+	if ( wp_is_post_revision( $post_id ) || 'tix_ticket' !== get_post_type( $post_id ) ) {
+		return;
+	}
+
+	// Stuff here is submittable via POST only.
+	if ( ! isset( $_POST['action'] ) || 'editpost' != $_POST['action'] ) {
+		return;
+	}
+
+	// Security check.
+	$nonce_action = 'update-post_' . $post_id;
+	check_admin_referer( $nonce_action );
+
+	if ( isset( $_POST['tix_type'] ) ) {
+		update_post_meta( $post_id, META_KEY, $_POST['tix_type'] );
+	}
```

Comment:
I would recommend using `filter_input()` here (everywhere) to retrieve the `$_POST` values, instead of referencing them directly. It gives you some built-in validation and sanitizing.

Reply:
The sanitization callback in `register_post_meta` takes care of this, but just to be extra-sure I added it.